### ISSUE TITLE
Piece heat maps discrepancies

### ISSUE THREAD

**Issue by ddobbelaere** [glinscott/nnue-pytorch#42]
I did some research on so-called "piece heat maps" (based on a hacked visualizer branch) and found something interesting. Define a piece heat map as the sum of all absolute values of all weights in the input layer that are connected to some (piece position, king position) pair.

Then we get the following heat maps for master/latest vdv on fishtest:

![master](https://user-images.githubusercontent.com/38748663/103476875-ab1c9e00-4db9-11eb-9957-f3186d38f751.png)
![vdv](https://user-images.githubusercontent.com/38748663/103476876-ace66180-4db9-11eb-9904-225939a7b73f.png)

Note the following two observations are much more outspoken for master:
1. Few "energy" for own piece where own king is located (left side).
2. Few "energy" for other pieces where own king can be put in check (only from adjacent squares for bishops, rooks and queens)

I suppose 2. is due to training data filtering (to yield only "quiet" positions where king is not in check)? But why is it so less outspoken on vdv net?

**Comment by ddobbelaere** [comment]
@hxim The code we use is based on the function downloadWeightsPng in https://github.com/hxim/Stockfish-Evaluation-Guide/blame/master/index.html

However, I had to change the `const inpos = {x:(7-kipos.x)+pipos.x*8,y:kipos.y+(7-pipos.y)*8};` to `const inpos = {x:kipos.x+pipos.x*8,y:kipos.y+pipos.y*8};` to make it work. Are you sure the code is correct?

**Comment by hxim** [comment]
> 
> 
> @hxim The code we use is based on the function downloadWeightsPng in https://github.com/hxim/Stockfish-Evaluation-Guide/blame/master/index.html
> 
> However, I had to change the `const inpos = {x:(7-kipos.x)+pipos.x*8,y:kipos.y+(7-pipos.y)*8};` to `const inpos = {x:kipos.x+pipos.x*8,y:kipos.y+pipos.y*8};` to "make it work" (i.e. look at the weights like they are logically arranged as a chess board with a8 in top left corner). Are you sure the code is correct?

It is flipped in the Stockfish-Evaluation-Guide on purpose. I flipped x and y positions inside 8x8 subblocks (king positions) so it is oppostite to as is on chessboard because it is more natural if you are looking on weights relative to king. In the non-flipped version (your version) the weights for piece on squares to the left from king are visualized on the right side of that 8x8 subblock (on the right of the "black dot") and the weights for piece on squares to the below king are visualized in top part of that 8x8 subblock (above the "black dot") and so on...
Also your formula depends on your interpretation of y coordinates which may be different (y axis can go from bottom to top or from top to the bottom)

**Comment by vondele** [comment]
again interesting. The fens used for training are different in the C++ and the python trainer. The C++ can do a qsearch, while the python one can't (easily with good performance). The python code will skip all fens where the bestmove is a capture or the king is in check.

**Comment by glinscott** [comment]
That is very interesting.  Was there ever an attempt to use the capture positions, just scoring them as the next quiet position in the game?

**Comment by vondele** [comment]
not exactly sure what you're asking, for the pytorch trainer I'm trying to describe '--smart-fen-skipping'. There have also been multiple attempts to retain only fens from a leaf in qsearch, or the last position along the PV prior to a capture.

**Comment by vondele** [comment]
@ddobbelaere you might want to have a look at the net of https://tests.stockfishchess.org/tests/view/5ff0b6f76019e097de3eec52
which is a slight change in the parameter used for training (ranger eps), but results in very different looking nets, of similar strength.
![image](https://user-images.githubusercontent.com/4202567/103463787-60564400-4d2f-11eb-9107-1a49ab03fd2e.png)

**Comment by ddobbelaere** [comment]
Oh wow, interesting. The mean input weight energy seems much higher also.

Just saying, the branch in #41 would help to avoid saturation (e.g. set --input-weights-vmin/vmax to +/-1).

**Comment by vondele** [comment]
yes, but the saturation is a property of the net as well. I guess you refer to it as 'weight energy', which probably is a better term.

**Comment by ddobbelaere** [comment]
Hmm, ok, but beware that all the yellow just means (in this case, i.e. with current master) that abs(w) >= 0.5, i.e. the "saturation" might be only a visual aspect.

**Comment by Sopel97** [comment]
The difference in color near the king may be explained by much lower virtual (factorized) weights. This *may* mean that the master net doesn't know piece values very well (briefly checking with https://hxim.github.io/Stockfish-Evaluation-Guide/ does indeed show that pieces have slightly weird values).

**Comment by ddobbelaere** [comment]
> The difference in color near the king (and overall higher smoothness in the vondele's net) may be explained by much lower virtual (factorized) weights.

Good point, I completely missed that (mainly due to my ignorance on factorization back then). It might be very interesting to also add support for HalfKP^ (non-coalesced) visualization.

---

@hxim We use the same orientation (visualizer.py and https://hxim.github.io/Stockfish-Evaluation-Guide/), i.e., y from top to bottom. Currently also the same formula is `const inpos = {x:(7-kipos.x)+pipos.x*8,y:kipos.y+(7-pipos.y)*8};` is used, but isn't `const inpos = {x:kipos.x+pipos.x*8,y:(7-kipos.y)+(7-pipos.y)*8};` much more logical still (flipping all the ranks because y is top to bottom)?

Using the latter formula (and y from top to bottom and x from right to left) we would get nice visual "chessboards" from the side to move perspective:

In a 8x8 subblock:

```
KA8 | KB8 | ... | KG8 | KH8
KA7 | KB7 | ... | KG7 | KG7
...
KA1 | KA2 | ... | KG1 | KH1
```

From piece superblocks (each item represents a 8x8 king subblock) perspective, example for queen:

```
QA8 | QB8 | ... | QG8 | QH8
QA7 | QB7 | ... | QG7 | QG7
...
QA1 | QA2 | ... | QG1 | QH1
```

Combined (taking into account pawns on first/last ranks are omitted), WP/BP stand for "white/black pawn" and so on:

```
KA8, WPA7 | KB8, WPA7 | ... | KG8, BPH7 | KH7, BPH7
KA7, WPA7 | KB7, WPA7 | ... | KG7, BPH7 | KG7, BPH7
...
KA1, WQA1 | KA2, WQA1 | ... | KG1, BQH1 | KH1, BQH1
```

Currently, I still fail to understand the logic of  `const inpos = {x:(7-kipos.x)+pipos.x*8,y:kipos.y+(7-pipos.y)*8};` used in both visualizer.py and https://hxim.github.io/Stockfish-Evaluation-Guide/.

**Comment by hxim** [comment]
@ddobbelaere
>  isn't `const inpos = {x:kipos.x+pipos.x*8,y:(7-kipos.y)+(7-pipos.y)*8};` much more logical still (flipping all the ranks because y is top to bottom)?

I will try to explain again why I think current/flipped formula is more logical. Look at this example:
With formula `const inpos = {x:(7-kipos.x)+pipos.x*8,y:kipos.y+(7-pipos.y)*8};` - flipped king coordinates you get this:
![obrazek](https://user-images.githubusercontent.com/6370548/103477106-9fca7200-4dbb-11eb-8ae5-eb5d2a9f50de.png)
In each 8x8 subblock you can see weights for white pawn on some fixed square/position on chessboard. But weights are different based on king position. As you can see two pixels above "black dot" (where king position equals white pawn position - imposible position - we always use black color) is weight always much higher (white color). If it was not flipped (using `const inpos = {x:kipos.x+pipos.x*8,y:(7-kipos.y)+(7-pipos.y)*8};` it would be always two pixels below "black dot". Now if you think about it what is more logical? Obviously having it above because this weight is applied when pawn is two squares above king.
So you want to use 
```
KH1 | KG1 | ... | KB2 | KA1
...
KH8 | KG8 | ... | KB8 | KA8
```
in a subblock instead of
```
KA8 | KB8 | ... | KG8 | KH8
KA7 | KB7 | ... | KG7 | KG7
...
KA1 | KB2 | ... | KG1 | KH1
```
This is the situation where weight is higher:
![obrazek](https://user-images.githubusercontent.com/6370548/103477391-d48bf880-4dbe-11eb-8da4-c5289332fb46.png)
This might be confusing because weights are actually for piece/pawn and not for king so you want weight for piece more to the right from the king on the right side of 8x8 subblock which means for lower x coordinate of king you want weight for it positioned in higher x coordinate of weight pixel (x is flipped). Same logic for y coordinate.

**Comment by ddobbelaere** [comment]
@hxim Thanks for the clear explanation, I understand now, makes sense!
